meteor/server/api/ingest/syncChangesToPartInstance.ts

```diff
@@ -179,3 +197,102 @@ export async function syncChangesToPartInstances(
 		}
 	}
 }
+
+/**
+ * Inserts given PartInstances and underlying Part to the list of PartInstances to be synced
+ */
+function insertToSyncedInstanceCandidates(
+	instances: SyncedInstance[],
+	cache: CacheForPlayout,
+	ingestCache: ReadOnlyIngestCacheWithoutRundown,
+	thisPartInstance: PartInstance,
+	previousPartInstance: PartInstance | undefined,
+	part: Part,
+	playStatus: PlayStatus
+): void {
+	instances.push({
+		existingPartInstance: thisPartInstance,
+		previousPartInstance: previousPartInstance,
+		playStatus,
+		newPart: part,
+		piecesThatMayBeActive: fetchPiecesThatMayBeActiveForPart(cache, ingestCache, part),
+	})
+}
+
+/**
+ * Finds the underlying Part for a given `thisPartInstance` and inserts it to the list of PartInstances to be synced.
+ * Doesn't do anything if it can't find the underlying Part in `cache`.
+ */
+function findPartAndInsertToSyncedInstanceCandidates(
+	instances: SyncedInstance[],
+	cache: CacheForPlayout,
+	ingestCache: ReadOnlyIngestCacheWithoutRundown,
+	thisPartInstance: PartInstance,
+	previousPartInstance: PartInstance | undefined,
+	playStatus: PlayStatus
+): void {
+	const newPart = cache.Parts.findOne(thisPartInstance.part._id)
+	if (!newPart) return
+
+	insertToSyncedInstanceCandidates(
+		instances,
+		cache,
+		ingestCache,
+		thisPartInstance,
+		previousPartInstance,
+		newPart,
+		playStatus
+	)
+}
+
+/**
+ * Finds the most recent Part before a given `currentPartInstance` within the current Segment. Then finds the
+ * PartInstance that matches that Part. Returns them if found or returns `null` if it can't find anything.
+ */
+function findLastUnorphanedPartInstanceInSegment(
+	cache: CacheForPlayout,
+	currentPartInstance: PartInstance
+): {
+	partInstance: PartInstance
+	part: Part
+} | null {
+	const previousParts = cache.Parts.findFetch(
```

Comment:
I don't think we need to be looking at the Parts for this.
The strategy I would use is to iterate backwards through the PartInstances (ordered by takeCount, and with a matching playlistActivationId), to find the first one without `orphaned: 'adlib'`.
Then do the same criteria you have below, but if the partinstance doesnt match then return null.

What you have here will find that PartInstance, but if its already been reset, or the part has been deleted in enps, it will find the one before. 
Is that actually desirable? I think this will lead to some confusing behaviour. 
Such as PartA has been modified in enps, but not synced because this is algorithm is finding PartB. PartB gets reset because they set it as next, and now PartA suddenly starts updating.

Reply:
Okay, how about this way? I guess this will be way more performant because it actually just requires a single find and sort through the PartInstances and then a single `_id` find in Parts.